syntax = "proto3";

package yandex.cloud.priv.compute.v1;

import "google/protobuf/timestamp.proto";
import "yandex/cloud/priv/compute/v1/validation.proto";

option go_package = "a.yandex-team.ru/cloud/bitbucket/private-api/yandex/cloud/priv/compute/v1;compute";
option java_outer_classname = "PP";

message PlacementGroup {
  // ID of the placement group.
  string id = 1;

  // ID of the folder that the placement group belongs to.
  string folder_id = 2;

  // Creation timestamp in [RFC3339](https://www.ietf.org/rfc/rfc3339.txt) text format.
  google.protobuf.Timestamp created_at = 3;

  // Name of the placement group.
  // The name is unique within the folder.
  string name = 4;

  // Description of the placement group.
  string description = 5;

  // Resource labels as `key:value` pairs.
  map<string, string> labels = 6;

  // Placement strategy.
  oneof placement_strategy {
    option (exactly_one) = true;
    SpreadPlacementStrategy spread_placement_strategy = 7;
    PartitionPlacementStrategy partition_placement_strategy = 8;
  }
}

message SpreadPlacementStrategy {
  int64 max_instances_per_fault_domain = 100 [(value) = "1-1000"];
  int64 max_instances_per_node = 101 [(value) = "1-100"];
  bool best_effort = 102;
}

message PartitionPlacementStrategy {
  int64 partitions = 1 [(value) = "2-5"];
  int64 max_instances_per_node = 100 [(value) = "1-100"];
}
